In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

## Ejemplo 7.1 Ecuación de salario por hora (wage1.dta)

In [2]:
dfej7_1= pd.read_stata('wage1.dta')
dfej7_1.head()

wage  educ  exper  tenure  nonwhite  female  married  numdep  smsa  \
0  3.10    11      2       0         0       1        0       2     1   
1  3.24    12     22       2         0       1        1       3     1   
2  3.00    11      2       0         0       0        0       2     0   
3  6.00     8     44      28         0       0        1       0     1   
4  5.30    12      7       2         0       0        1       1     0   

   northcen  ...  trcommpu  trade  services  profserv  profocc  clerocc  \
0         0  ...         0      0         0         0        0        0   
1         0  ...         0      0         1         0        0        0   
2         0  ...         0      1         0         0        0        0   
3         0  ...         0      0         0         0        0        1   
4         0  ...         0      0         0         0        0        0   

   servocc     lwage  expersq  tenursq  
0        0  1.131402        4        0  
1        1  1.175573      484        4  
2        0  1.098612        4        0  
3        0  1.791759     1936      784  
4        0  1.667707       49        4  

[5 rows x 24 columns]

In [3]:
regdfej7_1 = smf.ols('wage ~ female + educ + exper + tenure', data = dfej7_1).fit() 
print(regdfej7_1.summary())

c:\python\lib\site-packages\statsmodels\compat\pandas.py:49: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  data_klasses = (pandas.Series, pandas.DataFrame, pandas.Panel)


                            OLS Regression Results                            
Dep. Variable:                   wage   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.359
Method:                 Least Squares   F-statistic:                     74.40
Date:                Thu, 09 Sep 2021   Prob (F-statistic):           7.30e-50
Time:                        15:31:36   Log-Likelihood:                -1314.2
No. Observations:                 526   AIC:                             2638.
Df Residuals:                     521   BIC:                             2660.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.5679      0.725     -2.164      0.0

In [4]:
dfej7_1['wage'].loc[dfej7_1.female ==1].mean()

4.587659

In [5]:
dfej7_1['wage'].loc[dfej7_1.female ==0].mean()

7.099489

In [6]:
4.587659 - 7.099489


-2.51183

In [9]:
dfej7_1['wage'].mean()

5.896103

In [8]:
regA = smf.ols('wage ~ 1', data = dfej7_1).fit() 
print(regA.summary())

                            OLS Regression Results                            
Dep. Variable:                   wage   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 09 Sep 2021   Prob (F-statistic):                nan
Time:                        15:40:42   Log-Likelihood:                -1433.1
No. Observations:                 526   AIC:                             2868.
Df Residuals:                     525   BIC:                             2872.
Df Model:                           0                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      5.8961      0.161     36.616      0.0

c:\python\lib\site-packages\statsmodels\regression\linear_model.py:1554: RuntimeWarning: invalid value encountered in double_scalars
  return self.ess/self.df_model
c:\python\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
c:\python\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
c:\python\lib\site-packages\scipy\stats\_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


In [10]:
regB = smf.ols('wage ~ female', data = dfej7_1).fit() 
print(regB.summary())

                            OLS Regression Results                            
Dep. Variable:                   wage   R-squared:                       0.116
Model:                            OLS   Adj. R-squared:                  0.114
Method:                 Least Squares   F-statistic:                     68.54
Date:                Thu, 09 Sep 2021   Prob (F-statistic):           1.04e-15
Time:                        15:51:06   Log-Likelihood:                -1400.7
No. Observations:                 526   AIC:                             2805.
Df Residuals:                     524   BIC:                             2814.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0995      0.210     33.806      0.0

In [11]:
regC = smf.ols('wage ~ female + educ ', data = dfej7_1).fit() 
print(regC.summary())

                            OLS Regression Results                            
Dep. Variable:                   wage   R-squared:                       0.259
Model:                            OLS   Adj. R-squared:                  0.256
Method:                 Least Squares   F-statistic:                     91.32
Date:                Thu, 09 Sep 2021   Prob (F-statistic):           9.66e-35
Time:                        16:06:33   Log-Likelihood:                -1354.3
No. Observations:                 526   AIC:                             2715.
Df Residuals:                     523   BIC:                             2727.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.6228      0.673      0.926      0.3

## Ejemplo 7.2 Efectos de poseer un computador sobre la nota media en la univercidad (gpa1.dta)

In [7]:
dfej7_2= pd.read_stata('gpa1.dta')
dfej7_2.head()

age  soph  junior  senior  senior5  male  campus  business  engineer  \
0   21     0       0       1        0     0       0         1         0   
1   21     0       0       1        0     0       0         1         0   
2   20     0       1       0        0     0       0         1         0   
3   19     1       0       0        0     1       1         1         0   
4   20     0       1       0        0     0       0         1         0   

   colGPA  ...  greek  car  siblings  bgfriend  clubs  skipped  alcohol  \
0     3.0  ...      0    1         1         0      0      2.0      1.0   
1     3.4  ...      0    1         0         1      1      0.0      1.0   
2     3.0  ...      0    1         1         0      1      0.0      1.0   
3     3.5  ...      0    0         1         0      0      0.0      0.0   
4     3.6  ...      0    1         1         1      0      0.0      1.5   

   gradMI  fathcoll  mothcoll  
0       1         0         0  
1       1         1         1  
2       1         1         1  
3       0         0         0  
4       1         1         0  

[5 rows x 29 columns]

In [12]:
regdfej7_2 = smf.ols('colGPA ~ PC + hsGPA + ACT ', data = dfej7_2).fit() 
print(regdfej7_2.summary())

                            OLS Regression Results                            
Dep. Variable:                 colGPA   R-squared:                       0.219
Model:                            OLS   Adj. R-squared:                  0.202
Method:                 Least Squares   F-statistic:                     12.83
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           1.93e-07
Time:                        20:32:59   Log-Likelihood:                -42.796
No. Observations:                 141   AIC:                             93.59
Df Residuals:                     137   BIC:                             105.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.2635      0.333      3.793      0.0

## Ejemplo 7.3 Efectos de las subvenciones de formación sobre las horas de formación (jtrain.dta)

In [13]:
dfej7_3= pd.read_stata('jtrain.dta')
dfej7_3.head()

year     fcode  employ       sales   avgsal  scrap  rework  tothrs  union  \
0  1987  410032.0   100.0  47000000.0  35000.0    NaN     NaN    12.0      0   
1  1988  410032.0   131.0  43000000.0  37000.0    NaN     NaN     8.0      0   
2  1989  410032.0   123.0  49000000.0  39000.0    NaN     NaN     8.0      0   
3  1987  410440.0    12.0   1560000.0  10500.0    NaN     NaN    12.0      0   
4  1988  410440.0    13.0   1970000.0  11000.0    NaN     NaN    12.0      0   

   grant  ...  grant_1  clscrap  cgrant  clemploy   clsales    lavgsal  \
0      0  ...        0      NaN       0       NaN       NaN  10.463103   
1      0  ...        0      NaN       0  0.270027 -0.088949  10.518673   
2      0  ...        0      NaN       0 -0.063013  0.130621  10.571317   
3      0  ...        0      NaN       0       NaN       NaN   9.259130   
4      0  ...        0      NaN       0  0.080043  0.233347   9.305651   

   clavgsal  cgrant_1   chrsemp  clhrsemp  
0       NaN       NaN       NaN       NaN  
1  0.055570       0.0 -8.946565 -1.165385  
2  0.052644       0.0  0.198597  0.047832  
3       NaN       NaN       NaN       NaN  
4  0.046520       0.0  0.000000  0.000000  

[5 rows x 30 columns]

In [18]:
dfej7_3['grant'] = dfej7_3['grant'].astype(float)

In [22]:
# Se filtra la base para el año 1988
dfej7_388= dfej7_3[dfej7_3.year == 1988]

In [23]:
regdfej7_3 = smf.ols('hrsemp ~ grant + np.log(sales) + np.log(employ) ', data = dfej7_388).fit() 
print(regdfej7_3.summary())

                            OLS Regression Results                            
Dep. Variable:                 hrsemp   R-squared:                       0.237
Model:                            OLS   Adj. R-squared:                  0.214
Method:                 Least Squares   F-statistic:                     10.44
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           4.80e-06
Time:                        20:47:56   Log-Likelihood:                -482.29
No. Observations:                 105   AIC:                             972.6
Df Residuals:                     101   BIC:                             983.2
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         46.6651     43.412      1.

## Ejemplo 7.4 Regresión del precio de la vivienda (hprice1.dta)

In [24]:
dfej7_4= pd.read_stata('hprice1.dta')
dfej7_4.head()

price      assess  bdrms  lotsize  sqrft  colonial    lprice   lassess  \
0  300.0  349.100006      4   6126.0   2438         1  5.703783  5.855359   
1  370.0  351.500000      3   9903.0   2076         1  5.913503  5.862210   
2  191.0  217.699997      3   5200.0   1374         0  5.252274  5.383118   
3  195.0  231.800003      3   4600.0   1448         1  5.273000  5.445875   
4  373.0  319.100006      4   6095.0   2514         1  5.921578  5.765504   

   llotsize    lsqrft  
0  8.720297  7.798934  
1  9.200593  7.638198  
2  8.556414  7.225482  
3  8.433811  7.277938  
4  8.715224  7.829630

In [27]:
dfej7_4['bdrms'] = dfej7_4['bdrms'].astype(float)

In [29]:
dfej7_4['sqrft'] = dfej7_4['sqrft'].astype(float)

In [30]:
dfej7_4['colonial'] = dfej7_4['colonial'].astype(float)

In [31]:
dfej7_4.dtypes

price       float32
assess      float32
bdrms       float64
lotsize     float32
sqrft       float64
colonial    float64
lprice      float32
lassess     float32
llotsize    float32
lsqrft      float32
dtype: object

In [32]:
regdfej7_4 = smf.ols('np.log(price) ~ np.log(lotsize) + np.log(sqrft) + bdrms + colonial ', data = dfej7_4).fit() 
print(regdfej7_4.summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(price)   R-squared:                       0.649
Model:                            OLS   Adj. R-squared:                  0.632
Method:                 Least Squares   F-statistic:                     38.38
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           3.74e-18
Time:                        21:13:02   Log-Likelihood:                 26.619
No. Observations:                  88   AIC:                            -43.24
Df Residuals:                      83   BIC:                            -30.85
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          -1.3496      0.651     

## Ejemplo 7.5 Ecuaclón del logaritmo del salario por hora* (wage1.dta)

In [34]:
dfej7_1.dtypes

wage        float32
educ           int8
exper          int8
tenure         int8
nonwhite       int8
female         int8
married        int8
numdep         int8
smsa           int8
northcen       int8
south          int8
west           int8
construc       int8
ndurman        int8
trcommpu       int8
trade          int8
services       int8
profserv       int8
profocc        int8
clerocc        int8
servocc        int8
lwage       float32
expersq       int16
tenursq       int16
dtype: object

In [3]:
dfej7_1['educ'] = dfej7_1['educ'].astype(float)

In [4]:
dfej7_1['exper'] = dfej7_1['exper'].astype(float)

In [5]:
dfej7_1['tenure'] = dfej7_1['tenure'].astype(float)

In [6]:
dfej7_1['female'] = dfej7_1['female'].astype(float)

In [7]:
regdfej7_5 = smf.ols('np.log(wage) ~ female + educ + exper + np.square(exper) + tenure + np.square(tenure)', data = dfej7_1).fit() 
print(regdfej7_5.summary())

c:\python\lib\site-packages\statsmodels\compat\pandas.py:49: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  data_klasses = (pandas.Series, pandas.DataFrame, pandas.Panel)


                            OLS Regression Results                            
Dep. Variable:           np.log(wage)   R-squared:                       0.441
Model:                            OLS   Adj. R-squared:                  0.434
Method:                 Least Squares   F-statistic:                     68.18
Date:                Thu, 10 Sep 2020   Prob (F-statistic):           2.11e-62
Time:                        15:57:24   Log-Likelihood:                -260.59
No. Observations:                 526   AIC:                             535.2
Df Residuals:                     519   BIC:                             565.0
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept             0.4167      0.09

In [29]:
a = dfej7_1['lwage'].loc[dfej7_1['female']== 1 ].mean() - dfej7_1['lwage'].loc[dfej7_1['female']== 0 ].mean()

In [30]:
a

-0.39721763

In [23]:
np.exp(-0.297987699508667)-1

-0.25768952952487345

In [ ]:
# si se hace con fitted da igual

In [25]:
dfej7_1['lwagefit'] = regdfej7_5.fittedvalues

In [26]:
b = dfej7_1['lwagefit'].loc[dfej7_1['female']== 1 ].mean() - dfej7_1['lwagefit'].loc[dfej7_1['female']== 0 ].mean()
b

-0.3972174717365169

In [ ]:
# lo del libro es teniendo en cta las variables de la regresión

## Ejemplo 7.6 Ecuaclón del logaritmo del salario por hora* (wage1.dta)

In [44]:
dfej7_1['married'] = dfej7_1['married'].astype(float)

In [47]:
dfej7_1['marrmale']=(1-dfej7_1['female']) *dfej7_1['married']
dfej7_1['marrfem']=dfej7_1['female'] *dfej7_1['married']
dfej7_1['singfem']=dfej7_1['female'] *(1-dfej7_1['married'])
dfej7_1['singmale']= (1-dfej7_1['female'])*(1-dfej7_1['married'])

In [53]:
dfej7_1[['female','married','marrmale','marrfem','singfem','singmale']].head()

female  married  marrmale  marrfem  singfem  singmale
0     1.0      0.0       0.0      0.0      1.0       0.0
1     1.0      1.0       0.0      1.0      0.0       0.0
2     0.0      0.0       0.0      0.0      0.0       1.0
3     0.0      1.0       1.0      0.0      0.0       0.0
4     0.0      1.0       1.0      0.0      0.0       0.0

In [54]:
regdfej7_6 = smf.ols('np.log(wage) ~ marrmale+marrfem+singfem+educ+exper+np.square(exper)+tenure+np.square(tenure)',data = dfej7_1).fit() 
print(regdfej7_6.summary())

                            OLS Regression Results                            
Dep. Variable:           np.log(wage)   R-squared:                       0.461
Model:                            OLS   Adj. R-squared:                  0.453
Method:                 Least Squares   F-statistic:                     55.25
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           1.28e-64
Time:                        21:49:30   Log-Likelihood:                -250.96
No. Observations:                 526   AIC:                             519.9
Df Residuals:                     517   BIC:                             558.3
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept             0.3214      0.10

## Ejemplo 7.7 Efectos de la atracclón ffslca sobre los salarios (beauty.dta)

In [2]:
dfej7_7= pd.read_stata('beauty.dta')
dfej7_7.head()

wage     lwage  belavg  abvavg  exper  looks  union  goodhlth  black  \
0   5.73  1.745715       0       1     30      4      0         1      0   
1   4.28  1.453953       0       0     28      3      0         1      0   
2   7.96  2.074429       0       1     35      4      0         1      0   
3  11.57  2.448416       0       0     38      3      0         1      0   
4  11.42  2.435366       0       0     27      3      0         1      0   

   female  married  south  bigcity  smllcity  service  expersq  educ  
0       1        1      0        0         1        1      900    14  
1       1        1      1        0         1        0      784    12  
2       1        0      0        0         1        0     1225    10  
3       0        1      0        1         0        1     1444    16  
4       0        1      0        0         1        0      729    16

In [72]:
regdfej7_7 = smf.ols('np.log(wage) ~ belavg + abvavg + educ + exper + married + black', data = dfej7_7).fit() 
print(regdfej7_7.summary())

                            OLS Regression Results                            
Dep. Variable:           np.log(wage)   R-squared:                       0.233
Model:                            OLS   Adj. R-squared:                  0.229
Method:                 Least Squares   F-statistic:                     63.35
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           8.78e-69
Time:                        23:01:55   Log-Likelihood:                -965.22
No. Observations:                1260   AIC:                             1944.
Df Residuals:                    1253   BIC:                             1980.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3623      0.085      4.244      0.0

In [ ]:
# hacer la ecuación para hombres y para mujeres

In [3]:
dfej7_7M = dfej7_7[dfej7_7.female == 1]

In [4]:
regdfej7_7M = smf.ols('np.log(wage) ~ belavg + abvavg + educ + exper + married + black', data = dfej7_7M).fit() 
print(regdfej7_7M.summary())

c:\python\lib\site-packages\statsmodels\compat\pandas.py:49: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  data_klasses = (pandas.Series, pandas.DataFrame, pandas.Panel)


                            OLS Regression Results                            
Dep. Variable:           np.log(wage)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.197
Method:                 Least Squares   F-statistic:                     18.79
Date:                Wed, 15 Sep 2021   Prob (F-statistic):           1.94e-19
Time:                        15:57:53   Log-Likelihood:                -285.84
No. Observations:                 436   AIC:                             585.7
Df Residuals:                     429   BIC:                             614.2
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0820      0.130      0.631      0.5

## Ejemplo 7.8 Efectos de las clasificaclones de tas facultades de derecho sobre los salarios iniciales (lawsch85.dta)

In [73]:
dfej7_8= pd.read_stata('lawsch85.dta')
dfej7_8.head()

rank   salary     cost   LSAT   GPA  libvol  faculty    age  clsize  north  \
0   128  31400.0   8340.0  155.0  3.15   216.0     45.0   12.0   210.0      1   
1   104  33098.0   6980.0  160.0  3.50   256.0     44.0  113.0   190.0      0   
2    34  32870.0  16370.0  155.0  3.25   424.0     78.0  134.0   270.0      0   
3    49  35000.0  17566.0  157.0  3.20   329.0    136.0   89.0   277.0      0   
4    95  33606.0   8350.0  162.0  3.38   332.0     56.0   70.0   150.0      0   

   ...  east  west    lsalary   studfac  top10  r11_25  r26_40  r41_60  \
0  ...     0     0  10.354563  4.666667      0       0       0       0   
1  ...     0     0  10.407228  4.318182      0       0       0       0   
2  ...     1     0  10.400315  3.461539      0       0       1       0   
3  ...     1     0  10.463103  2.036765      0       0       0       1   
4  ...     0     1  10.422460  2.678571      0       0       0       0   

    llibvol     lcost  
0  5.375278  9.028818  
1  5.545177  8.850804  
2  6.049734  9.703206  
3  5.796058  9.773721  
4  5.805135  9.030017  

[5 rows x 21 columns]

In [79]:
dfej7_8['r61_100'] = np.where(((dfej7_8['rank'] > 60) & (dfej7_8['rank'] <= 100)), 1, 0)   

In [80]:
dfej7_8[['rank','top10','r11_25','r26_40','r41_60', 'r61_100']].head()

rank  top10  r11_25  r26_40  r41_60  r61_100
0   128      0       0       0       0        0
1   104      0       0       0       0        0
2    34      0       0       1       0        0
3    49      0       0       0       1        0
4    95      0       0       0       0        1

In [81]:
regdfej7_8 = smf.ols('np.log(salary) ~ top10 + r11_25 + r26_40 + r41_60 +r61_100 + LSAT+GPA+ np.log(libvol)+np.log(cost)', data = dfej7_8).fit() 
print(regdfej7_8.summary())

                            OLS Regression Results                            
Dep. Variable:         np.log(salary)   R-squared:                       0.911
Model:                            OLS   Adj. R-squared:                  0.905
Method:                 Least Squares   F-statistic:                     143.2
Date:                Mon, 13 Jul 2020   Prob (F-statistic):           9.45e-62
Time:                        23:26:37   Log-Likelihood:                 146.45
No. Observations:                 136   AIC:                            -272.9
Df Residuals:                     126   BIC:                            -243.8
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          9.1653      0.411     22.

## Ejemplo 7.9 Efectos del uso del computador sobre los salarios ()

## Ejemplo 7.10 Ecuación del logaritmo del salario por hora* (wage1.dta)

In [3]:
dfej7_10= pd.read_stata('wage1.dta')
dfej7_10.head()

wage  educ  exper  tenure  nonwhite  female  married  numdep  smsa  \
0  3.10    11      2       0         0       1        0       2     1   
1  3.24    12     22       2         0       1        1       3     1   
2  3.00    11      2       0         0       0        0       2     0   
3  6.00     8     44      28         0       0        1       0     1   
4  5.30    12      7       2         0       0        1       1     0   

   northcen  ...  trcommpu  trade  services  profserv  profocc  clerocc  \
0         0  ...         0      0         0         0        0        0   
1         0  ...         0      0         1         0        0        0   
2         0  ...         0      1         0         0        0        0   
3         0  ...         0      0         0         0        0        1   
4         0  ...         0      0         0         0        0        0   

   servocc     lwage  expersq  tenursq  
0        0  1.131402        4        0  
1        1  1.175573      484        4  
2        0  1.098612        4        0  
3        0  1.791759     1936      784  
4        0  1.667707       49        4  

[5 rows x 24 columns]

In [13]:
# previamente se pasó a float en ejercicio 7.5
femeduc = dfej7_1['female']* dfej7_1['educ']

In [14]:
regdfej7_10 = smf.ols('np.log(wage) ~ female +  educ + femeduc +exper + np.square(exper) + tenure + np.square(tenure)', data = dfej7_1).fit() 
print(regdfej7_10.summary())

                            OLS Regression Results                            
Dep. Variable:           np.log(wage)   R-squared:                       0.441
Model:                            OLS   Adj. R-squared:                  0.433
Method:                 Least Squares   F-statistic:                     58.37
Date:                Tue, 14 Jul 2020   Prob (F-statistic):           1.67e-61
Time:                        12:07:20   Log-Likelihood:                -260.49
No. Observations:                 526   AIC:                             537.0
Df Residuals:                     518   BIC:                             571.1
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept             0.3888      0.11

## Ejemplo 7.11 Efectos de la raza sobre los salaries de los jugadores de béisbol (mlb1.dta)

In [15]:
dfej7_11= pd.read_stata('mlb1.dta')
dfej7_11.head()

salary     teamsal  nl  years  games  atbats  runs  hits  doubles  \
0  6329213.0  38407380.0   1     12   1705    6705  1076  1939      320   
1  3375000.0  38407380.0   1      8    918    3333   407   863      156   
2  3100000.0  38407380.0   1      5    751    2807   370   840      148   
3  2900000.0  38407380.0   1      8   1056    3337   405   816      143   
4  1650000.0  38407380.0   1     12   1196    3603   437   928       19   

   triples  ...     runsyr   percwhte   percblck   perchisp     blckpb  \
0       67  ...  89.666664  70.277969  18.844229  10.877804   0.000000   
1       38  ...  50.875000  70.277969  18.844229  10.877804  18.844229   
2       18  ...  74.000000  70.277969  18.844229  10.877804   0.000000   
3       18  ...  50.625000  70.277969  18.844229  10.877804   0.000000   
4       16  ...  36.416668  70.277969  18.844229  10.877804  18.844229   

   hispph     whtepw     blckph  hisppb    lsalary  
0     0.0  70.277969   0.000000     0.0  15.660686  
1     0.0   0.000000  10.877804     0.0  15.031906  
2     0.0  70.277969   0.000000     0.0  14.946913  
3     0.0  70.277969   0.000000     0.0  14.880221  
4     0.0   0.000000  10.877804     0.0  14.316286  

[5 rows x 47 columns]

In [17]:
dfej7_11.dtypes

salary      float32
teamsal     float32
nl             int8
years          int8
games         int16
atbats        int16
runs          int16
hits          int16
doubles       int16
triples       int16
hruns         int16
rbis          int16
bavg        float32
bb            int16
so            int16
sbases        int16
fldperc       int16
frstbase       int8
scndbase       int8
shrtstop       int8
thrdbase       int8
outfield       int8
catcher        int8
yrsallst       int8
hispan         int8
black          int8
whitepop    float32
blackpop    float32
hisppop     float32
pcinc         int16
gamesyr     float32
hrunsyr     float32
atbatsyr    float32
allstar     float32
slugavg     float32
rbisyr      float32
sbasesyr    float32
runsyr      float32
percwhte    float32
percblck    float32
perchisp    float32
blckpb      float32
hispph      float32
whtepw      float32
blckph      float32
hisppb      float32
lsalary     float32
dtype: object

In [18]:
dfej7_11['years'] = dfej7_11['years'].astype(float)

In [19]:
dfej7_11['black'] = dfej7_11['black'].astype(float)

In [20]:
dfej7_11['hispan'] = dfej7_11['hispan'].astype(float)

In [21]:
blaper = dfej7_11['black']*dfej7_11['percblck']

In [22]:
hisper = dfej7_11['hispan']*dfej7_11['perchisp']

In [23]:
regdfej7_11 = smf.ols('np.log(salary) ~ years +  gamesyr + bavg + hrunsyr + rbisyr + runsyr + fldperc +allstar+ black + hispan + blaper + hisper ', data = dfej7_11).fit() 
print(regdfej7_11.summary())

                            OLS Regression Results                            
Dep. Variable:         np.log(salary)   R-squared:                       0.638
Model:                            OLS   Adj. R-squared:                  0.624
Method:                 Least Squares   F-statistic:                     46.48
Date:                Tue, 14 Jul 2020   Prob (F-statistic):           1.31e-62
Time:                        14:46:04   Log-Likelihood:                -350.12
No. Observations:                 330   AIC:                             726.2
Df Residuals:                     317   BIC:                             775.6
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     10.3437      2.183      4.739      0.0

## Ejemplo 7.12 Un modelo lineal de probabildad de las detenciones (crime1.dta)

In [24]:
dfej7_12= pd.read_stata('crime1.dta')
dfej7_12.head()

narr86  nfarr86  nparr86  pcnv     avgsen    tottime  ptime86  qemp86  \
0       0        0        0  0.38  17.600000  35.200001       12     0.0   
1       2        2        0  0.44   0.000000   0.000000        0     1.0   
2       1        1        0  0.33  22.799999  22.799999        0     0.0   
3       2        2        1  0.25   0.000000   0.000000        5     2.0   
4       1        1        0  0.00   0.000000   0.000000        0     2.0   

   inc86  durat  black  hispan  born60  pcnvsq  pt86sq    inc86sq  
0    0.0    0.0      0       0       1  0.1444     144   0.000000  
1    0.8    0.0      0       1       0  0.1936       0   0.640000  
2    0.0   11.0      1       0       1  0.1089       0   0.000000  
3    8.8    0.0      0       1       1  0.0625      25  77.440002  
4    8.1    1.0      0       0       0  0.0000       0  65.610008

In [ ]:
# Primero creamos la dummy arr86 con series arr86 = 0. Luego restringimos el sampl con narr86 > 0 . 
# Lo creamos fuera del dataframe, pues no hace falta. En la regresión puedo poner series separadas

In [32]:
arr86= np.where(dfej7_12['narr86'] >0 , 1, 0 )

In [ ]:
dfej7_12['arr86']=np.where(dfej7_12['narr86'] >0 , 1, 0 )

In [27]:
dfej7_12['narr86'] = dfej7_12['narr86'].astype(float)

In [28]:
dfej7_12['ptime86'] = dfej7_12['ptime86'].astype(float)

In [30]:
dfej7_12.dtypes

narr86     float64
nfarr86       int8
nparr86       int8
pcnv       float32
avgsen     float32
tottime    float32
ptime86    float64
qemp86     float32
inc86      float32
durat      float32
black         int8
hispan        int8
born60        int8
pcnvsq     float32
pt86sq       int16
inc86sq    float32
dtype: object

In [33]:
regdfej7_12 = smf.ols('arr86 ~ pcnv +  avgsen + tottime + ptime86 + qemp86', data = dfej7_12).fit() 
print(regdfej7_12.summary())

                            OLS Regression Results                            
Dep. Variable:                  arr86   R-squared:                       0.047
Model:                            OLS   Adj. R-squared:                  0.046
Method:                 Least Squares   F-statistic:                     27.03
Date:                Tue, 14 Jul 2020   Prob (F-statistic):           9.09e-27
Time:                        16:33:35   Log-Likelihood:                -1609.7
No. Observations:                2725   AIC:                             3231.
Df Residuals:                    2719   BIC:                             3267.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4406      0.017     25.568      0.0